### PR TITLE
feat: support daisyui v4

### PR DESCRIPTION
**Pull request by ameinhardt**
summing up changes from #38 by @LeekJay with minimal changes

### CODE REVIEW

**Comment by SpasiboKojima** [comment]
Thank you for compiling it together, works for me so far with daisyui@4.6.0, hopefully @kidonng will take a look and merge it.

File: index.ts [hunk 1]
```diff
@@ -32,18 +30,44 @@ const defaultOptions = {
...
+ */
+const notAtruleNode = (node: ChildNode): Rule[] => {
+	const collectedRules: Rule[] = []
+
+	/**
+	 * Recursively collects non-atrule nodes.
+	 * @param {ChildNode} currentNode - The current node to examine.
+	 */
+	const collectNonAtrules = (currentNode: ChildNode) => {
+		if (currentNode.type === 'atrule') {
```
Comment (OEvgeny) [inline index.ts]: I had to do the following change to make TSC happy:
```suggestion
		if (currentNode.type === 'atrule' && currentNode.nodes) {
```

**Comment by OEvgeny** [comment]
Just for everyone using this with Daisy v4. Looks like this has an issue with button hover handling, probably because the piece of styling is wrapped in both `@media` and `@supports`.

I was able to work around it by adding the following styles manually:

```css
@media (hover: hover) {
  @supports (color:color-mix(in oklab,black,black)) {
    .btn:hover {
      background-color: color-mix(in oklab,oklch(var(--btn-color,var(--b2))/var(--tw-bg-opacity,1)) 90%,black);
      border-color: color-mix(in oklab,oklch(var(--btn-color,var(--b2))/var(--tw-border-opacity,1)) 90%,black);
    }
  }
}
```

**Comment by skf-funzt** [comment]
@ameinhardt @kidonng 
Will this PR progress soon?

**Comment by skf-funzt** [comment]
> Just for everyone using this with Daisy v4. Looks like this has an issue with button hover handling, probably because the piece of styling is wrapped in both `@media` and `@supports`.
> 
> I was able to work around it by adding the following styles manually:
> 
> ```css
> @media (hover: hover) {
>   @supports (color:color-mix(in oklab,black,black)) {
>     .btn:hover {
>       background-color: color-mix(in oklab,oklch(var(--btn-color,var(--b2))/var(--tw-bg-opacity,1)) 90%,black);
>       border-color: color-mix(in oklab,oklch(var(--btn-color,var(--b2))/var(--tw-border-opacity,1)) 90%,black);
>     }
>   }
> }
> ```

For Astro one needs to add this at the very top file possible as global, unscoped CSS style:

```astro
<style is:global>
  ... The mentioned workaround here ...
</style>
```

**Comment by s3xysteak** [comment]
Any news?

**Comment by mklueh** [comment]
@ameinhardt can you say a few words about the state of this repo / pull request? 
Seems like you've created a daisyui 4 compatible project as an alternative https://github.com/ameinhardt/unocss-preset-daisy

Can we safely use yours instead now? 😊

**Comment by ameinhardt** [comment]
I don't know about this repository.
My alternative not only checks the first token of each css rule and yields the full rule as string, but clusters the parent selector, full selector and body by every class name. With recent https://github.com/unocss/unocss/pull/4308 , I can keep the order of the rules while unocss magic applies any variant.
[Try it in the playground](https://unocss.dev/play/#html=DwEwlgbgBAxgNgQwM5ILwCIC2CAeBaAdzyRgCcBTcgOz0xCk3wQFcAXAeygAc8AWdAHwAoKFFCRYiFBgBmccjihhW5TEjwxqK0oJGixACwCMk5GnQqcrPAFYccKDPZVrAZgAM7hqT6CAqlTsAMIAyiFQANRQIAhgSACefgCSwAD0xsL6YghQBhQyGAasrFxIAFypqQDmygbMAEYAdDDsmKkImORgVAYIpCCsqcyBMCh4XBRI5NYxcfG6WaLi0PBmGGB47DCsYC002KQA1ng1rHX1eEYAbFCWbva57BDkpGXsXAgwyvG27oJp4AgmX0aQQwIBkGEemABgATKZpFgfvC7nhXPZBABBbZgZxINJwqFLAyuBHmJwuPD1dhweiYH4mVEYgQAITYHCoBNc4MBZIw9L4UCQH00eB%2B-GBS0BkpBBl4fPQFOs1NpgjZxTxUAItSg9VICCo9BacHYpHx6V4MqW9XZzgV9VYVDVts5qRtGqoVrE7o59sdusdeCo5DY%2BrgggAciHWGG0j7nF7gPGqH6Uw6aBMwAd5gIAAqkLN9eJxl2J5OpgM0KZ7GKkHMhcg1oslj1ll0V9N4T6aFxYmA91gtjltj0dwNVAzsJCsQQAcUn06HCb0WST7dWiPTlbwcG6h0EABk90vPSuxKlpWflom5QqlVSaSBnR6kFqddOECpYDTTea5SPfQ3cwt07bonEEJIqCcE8ALtICMBAwMkGYftyBQQQQhQzQUBgs9rXXKRgP9TsCD6KhuiqQQAHUyIo3DFm9Ai1nQRCaBeUhTUEABRUgONIeiQQvSEr0vBiYXleDFWcZVH0EAB5Nhd2DXUXT-S08MY0dJNYzZFO6chBAAEXIGQWDgQc3VLDS1y0wiEOIwN2D05TO0zbNBHzQs6wE1dy20hyaCc1glPIbdq2cWt60bCLm0s1trL8uyWIC3Tgv07duy0PsBx889RJBfLV1vST7xVJ9WXbJAwAALzQgl1LExLmJ0uBKIEA8%2Biqchcvw2zmsdSNTWwOAes0wCkp0pBMAw4aRri4cEqYzcUpwJBBAAFW6Yt5uXVchKBEThLE4qktK2SBAAJTQrg8UgULyHkToXDU2CU38lNYXsMpXILdyBE%2Bubk1eisvp%2Bryc3sUabPG-qU1a77AzcotBFaqGmuWlM6ARjNfuRgQ6DRpaiJTKbsfGXG60EKbcohA6ln2okxF5SSBQlK8STvaSH1pBgGVuBRrGZQyOK4EB2AIV0SRlZYFRAEWxYlhZFmARB6ge24EHqboQAUDA-grTBLkEIJdxgQ40lVh7E2YOANNEVhNe13X0D%2BO3Zfl8WaD2FQXAYahmG4PB4SQXoFd1KoqWQchLk8KAOOGHWQAfRQiBsWF0A0r0ll3ARgDBJIVEwKAjFBXPUhzt2VbAXP88LqBYVLi3q%2BsoY4GlhnoQ76FmaS1mleJUkSq5srecufmrDwZkAFl2BiOapavS2HCcUh%2BR%2BTBZ4QcMDc3uAqRdQR3moBhd4tzWrc7y9u4kFnxX7wxB9O4fH1HxkBcn8MBBCUiuC5aWl4VEgH%2B98ljdC4GwW48QuDkAwDAAwjZDjUhwOgKAqREw92YkAhAPBnDyQjLTdBN8kpYJwTIGQ8kABiFCCFXnLufNul9hLQjhAqAUKJ37ok-oZT8ORwDCkQNtQkQhr4rF7nfaWHMh6UhHgKN%2BE9mSYnkKQCyC9VwYMRFvF4rAoCaOUcQLCaFXwhwQArHcbVrKFUWIAUHJK7CgNAIAAmk5Ug3BmCkDgVHXIyBdSUDes4GQYBSCdBAAAQjSHY08YkO57XyjQkRrDxHs0fsxM6PNZHj0Fp-FkJiup-0OqI5qOSDICB1qZG2FlLEy20kU3URSgzRjDIIYMoYt40LUUQwpIAuq1K6dHJGlMBD9O2pU9RwEan1DqeFQ0eMpmRTaQVDpm5xl1Myr2AQqyKlHQWQUpZvSeldROAuGcAgJxTk2XTPKTCpSLPMH3CRyTESpLpHzJkn8gh9BAHk9pOzzAwA%2BVqPAABOG49QI4TKmDHLwxjTHMmsgEqobjyBZzEFmKobtRBIHcYUYopQKipFRY0WYCRmBgGaK0fF2AupIFSNOLYhxUhcEnBwS4Vx3BXCMO4AA7DYGwnKjCApsHgdw5BYSAoQLweoNh6i8AQAgRoBByD1C4BnBiWQt6sAwCEScaEVWqrQS3eFiLCE-NgR8h8IAcyV0kU-aRL8-n9DwDsYKxStXsDQmE9I3JK5cAEEkGQOjojsCqLAeBBAjHatfAQBcoUQ5utfGLNCuRQpwPYFOcgAB%2BNIPrK6jNNQ6z4OxNQACtmDTjADIH41ByrorGnBCaKUhnOniFACM4tRr00sR2rZXaLlxOuSapEvh7mc1tWkl579mRBBpIgUo3VPXS0BBpB2WtDTO1dgxSSxoZ1TG-HAbd0c%2BgcQILqU0OsXHUn6C8SOEKPBeFBde6OHKvDx1XUnJBurRDGoVFu7BEKnXyAyR-RwXNglgGYNNZFFCtilr9rcTgUxk3SS0C3TtTMbmwOnb%2B6O3ssrIuAD65dTscB6xQZ%2BGMYB3ShTiFAYM2EkBFjgwwBAhxQpnFCryJwMBS2a3kFmxM0TBKxK7v2hJQ6kkjpkmOserzjZORcArSWXrvkKiqAWegqmwBJzkOLDQNIoBVGwbYQDA4Xg6LYOwXTcB1DYGQcat2JV5CKDkAoSzgd4T3uaTGLecc5OJ2ToBzzYZdMuCypnWxHw3pJRaMMVgCngOUg3oEQDdhP41uABEoUrB4jyAwHgPAEAt7MHIGUKAvB05lwiXhmlEXkUxHiEgFD3btn2dOo5xwjnXM8HcxHQL3mX1%2BaQQFhpW9gs%2BxnGF1VYgMubrk7Fz28XrCJc4KiFLIDlYZenNlmB6A8sFbgEVkrRhyvhJq5Xar9i3aTjcQ1qJqHlgtZSW15z%2BBjRufDvUlpDh%2BvkDfewRQqJet7xw72Cbqr0sRe-bNuL94lvJdhZNpYG2ss5Z2-lwrxWoCuAlCdi7k2ceRNVZgbojWLnfIe48p7HXXtdfe4DnzCcfv%2BYB8NoHSGQcMTw9NqLUP5sw%2BcMt9%2Bq3kWI4h5tlHu30clZsFcf452CdRMq27asJP25CaviJ2%2BYmioPPJM-KTcjMmCAANL1E%2BfOq8iCjRJUt4IQAGMRpEt9LS3CobcCEAKTEDvTdO9Ny703ghADnhJ7kA3urfMVd4AYGIg-gmE2hgddzxNSMk886TE7P75iDZMP8Sntl3kpy517BmutrfwxxVThiFQTAz%2BX1OMuoB7aKyRpjxH0BPtl5XsvKBEzt8zxX0vPea8oPr9tjlKCbMYFb2Xbvhiu99-L5JKfKAAXS8H%2BjjAvB9Zj5b54Nvs-O-WQX6%2Befu-I22Fr0PjAnKN%2B4HH9vyfx%2BZ9V8X0fx-J-l919X1vq-zeJ9Zvv7QtXVyhg8Imu7CE8nCRkPCUAkE4CKisIjMVSYiWusoOusgeuyeBuH8xs8CZsSCXyOeQ8QSo2HEqWYkACkkZB12po4w7A3Q2ga2IuBoCoS8jqAsggV0nQmAasLinQ%2BOeGYCbANaWW0CsC2BiCf2H6qqcCCCP2Ih0hT4Nam6ohg2UhOBf25M4MEhWQ%2BqUSS8quWyfaseombM2uEm3M6BgGzI60AsSgVAMBeBGuT8hB3sxBAKMgNsDgNmhAk8a0iYZBSUS89BU2EO5B9CLBVg1EvQ2i1G8QziNGHQGavB4WjBIRase8qI6qWI5kUAS8iROh9CiY-BrAbsQh22dwmh%2BgXAiAmgk4tILwGA60UCoU8CFA5RogkkhRNhMBD4l6FA9ARAbhe6TeXhq05R2hysfhzEARnOwR-hoR6R5kmR2iORcu0xyRsxqRYR1gGRAgiiSx9CuRysdCqRehvaMeCBzE8eJhieZhr8Fhn8F0JiuI9hhhBBhszhNIxeExiIFBZoVBN0tBLwgR4OaxkxcxrBl0P23AYAe6BxYkhRghjRGA%2Bo4A7ArR%2BgVA8RSJjxFmMuChSUyJuIIach30EcvRvw66COKhLGIAbsYxe0uh-%2BTW5xjyQ0RBHxvh9CTBnJXGvxPg-xIWOgqxkWoJGxZRAg9Q%2B2oUXA0Jc0iueO8JCOkCwh6ABJqJNaGJnQWJKJeAuJipkkqpRJZsP2JJVIkp5JaJbRohP2tJ-GDJMS%2BhZxLCIBgG4B7UCAMRbABIcBwiDhFxiSVxNqSetxMmAgU8yA5s5uq4qKGkLM4ZTGSAxwSAAAjiSu4vIOUZijANiiUOUJUASkSvECSmSm0IWFSjShwGbAykyhZkYNLpylcDYFjkYAABzNn1m2D1D1DCrS5XDkBGAwCwjyqKrKpnhjEGEwjAGIGgH3CfwRgIAQBgAGaFqSzenxKa7GHIGmEyLjryJZIUAmJkDgb1BZ4LroYWDvxTS6j7kgCHmcE%2BHWQ2x4Y5x5wCAAASrQc6YITcT51cL5hk0GT0rA5oX55czceOOcmIIA9A-5XGgF35Lcj5jJpx6uLxiBG5A8W5L86SIZR4VAEZqiIIOQ5Be42Re4eAk4zwgpSQAA5EXDkFVJgJUaFEpBGWCIwgdGuWhffDCCgVJKOuYSGVPP7M8cADbKwv7O9uCtHLCLHDXvTq%2BsnJ8b%2BbXKoMXI3KBb4cpb6nXA3O0GXBXKQVpQXKpa4OpQZYJIhT2vAbmoOuhQ-JhfrncZGPORMvxJGfgUlBiRAK5ZJVHJCvJQNn9kKKHDpvDsrDZc9kbHhkRfWmmOOEcjRkNCNn8juiGQWckKXCrhYmeZFYEMGECWJSzBJU9MwORaaDVNJN5lwPgEYIEdnFpbhaxfpWBQjlXMLgVNMLEFZu1auMhJgNmHmH0FoOEuBgNTWquIVUlDTvelJZCnVWDs%2BWCI1WpXpfBYqYZTXO1CRbpSBeZeteeJZetQCJ1dCUgO1WtWDq3PxpUgJuOTZZcZudcduSnruR5AgDUBicuSJTZUWjQU6MDJJL9d0HgMoKpemIICXDtPLiCOjOYEDTQKDUXDpAWndIILpUDItH1IiPDSDXXODQIKZVDQDUlDjYjZWIILwDTLdWcfdf6Y9YGTcdhanhhCoKUCJZNZgqzfeYZYAqzZluQDwI2hCTUNOC8BdauLuLzQLfzYLRTDmEEJOGmtwIgK6HtSCJLZJKLcqnmG4h4lMOLerWAFLdrVdJoHdFAOniAChBZGrWkIdXdWeQ9RhU9VhTuYbgIOtJrN9WecutzcrDFcxMuhrBcBemej9htJrCtWxWJAHYiEHcut0WHfQAnSjc8BHfUPXJlTHQqPHZrInS8OHR7ZHYTdHYJg6ShROaJtOWiNPHSswL-OkKub6YiE7fZS7Y5SGVOjrN7XHnXTwC0DrMXhMOxjwuTCZGAM3gACT-AD1Iq2FFxG3DBbBjDD1TAzCxAJBpCz135Ir74UDRCj3D0BLN4CAoKSSoikSkDkRUBtTADb3dAfh7oUSNAv1b2zxIq-671iTD0H0Oxj3H0YCn050XkGLoS5zb3-nBgerb2f0nHR4V201IHO0M3PUYHMi5iTjBg92sJ93jCYPFLZUDp-KdD6iy53Y2V8KVHulAk2V9CsDUj-J0MMMOo6wbzcD4NRWvmkrzL2mk6XLIWAHMm3J03IMpJoHBnM0CBUTaxtruXN23K4PaiGjiwno9G%2BU3q36EO54KBQAlploVoaBaCmbVWChcAMggpgp%2BUyV-BvkPQmgeo3UAEcVLAhJ5ZQAzxzxQB5bggdElGyGqHIJKAgBry0C7xn29y7yOpBpVDpmoI8iO1hOnm91zz%2Bbfa-a2bWTWpiP8WAatSyY336isA2yfi4hUCvj6jKNFyQTaDBjaKlovAeoEWrhcAV4iOLBOLMDUXPA%2BLHxTDyDbCQkrwBoeJUCaCMZdTaLOChTxDkB9BQDsD%2BrITHlkBgBcDLmMb1NuzSOHCrM-ZgA5BDMyAUDkAhItzZpiT3WRMo24JPmckrwhMbxzzhOwyCAOLukep2mHGOPl3CQ0JAA&config=JYWwDg9gTgLgBAbzgEwKYDNgDtUGEJaYDmANHGFKgM6owCCMMUwARgK4zDoCeZF1tAJIBjAlT6UaMAKpYIcAL5x0UCCDgByNnOFUqGgFChIsROUm0AIgENgVbouWr1GgALWQqbAAtrUZDAA9NoQulQAtPxS4ci29oYGqAAeJvBo6NZsADZpGNh4BMQAFAgGcOYCMFQAXHAA2mXlFVKyEEUAlCSN5VG0DEysHFzcHV1NzUKiWFQl3U1UwtZZqLUAjAB0AExj43DCyFi1Gt6MYDWBgdQg61TegRo75Qqdc70wNnYjL%2BUAul3PBiAA&css=AIWwpgJglghgBACgA4CcwDMwoM4FoDGA9gDaEq7b4AWY4AXHBDCgNYCUcA3gFBxx0pChAC5defOEVLlKNeo2YsA3OL65cUAHYMA7ACYAdAAYAbAFI4RgwEYAnNbh6AzNYMmVEuOuwBXBiYAWAwAOCytrAFY4axMjDwl1AHcYBmCgnWswm1tbOGCnEJ14tVwsXVcdc0sDPT0HWptg4q9cJHwGawLO0OqjJ1s9Rx0Ig2Hm73booNsqqyMIwacAwwiI8dwYSetpoz0svqXo-Ldrda1JozDLdd8Lq7jVFuS7y2vH9SwXozfPdUEfTQQSC4ABGhAAHh00CB1v9AcCQcJtNUItDYYQAUCIKCYBAAOZgDoGWxo94bTRQEAwYRQQiaUFIhhWPQRbDrGAUqk0ukaTRIHzCJk1NlkxH09CEfA%2BPCUGDEQnVWxrUVkIHkMUdJDg9bCGAg0GqsrRLU6vW4FC4qDSoWo2jrJD%2BEYmWxZOwmIYjMZk7C6II9ZnupYrZW-Da%2BgwRTLVGI6I4FEynMnIlwGJzBSL7OrWIzBRwRIKdexe0OaSY6IKBHQ6TMxRyBAwBJwRGYBdYg6wMFmpvQmau9SLBGZ5wwBazDExNUV6TsFHtGPtzbrzvNdIymYsJUFOTuubZpfbWGJRFlBHM6Wz9NtlisBKs1909oKN5uBeIAX24H%2B4AG0mLrcMIchgAAvAA5MQUB4lQwigQAumIfBSGQFDULQCoQVBwjxOoWi6IYpiuvYjgpu44jeH4cD1v6NhRDEDwlMkqTpFG4Q5HkBTBEUZGlCg5SjLMNR1I4hjWJOJRtKktghHoAT7AmAweqMIYtJsHQjE2ey9E4E5xic2G8l8PwtLcTL3Ppzyma89EtJ8lnfNZfwYvC2JgpC0SkiUcJYgyyJWLaMLcV5CK4gSRIkna3EcpS1K0vSGrVCyIolFFXKxby-KCglSUtGKuASlKMr4HKCpWEq%2Blgigao%2BZq2rcbq%2BoVWqNX6fV5qWtaKIeS0DpwAEUnWJpVhOEYQ56MMSn6T6lFSRkWROHouRBhGykTAwtjBKMLHGNmgbLMt%2BkpHAlT8a6aS6Qm%2BnJiJOjBBEma3g%2BqyjEYMQyZdkyDg2OSyf2A7HgsxgzGOpElO2HRrmZ3EgtOcAzDYkQsUZ6ggtusMjjELq9C9ObRLYOg2E4wzNuVkxjSE87UbsSqjREJghLeSp6CoH5AA&options=N4XyA)
(daisy color's are not properly converted via esm.sh in the browser, that's why I added them again here)

**Comment by mklueh** [comment]
@ameinhardt great! thanks man, works perfectly 😊✌️

**Pull request closed without merging**